sdk/eventhub/Azure.Messaging.EventHubs/design/proposal-custom-endpoint-address.md

```diff
@@ -0,0 +1,69 @@
+
+## Terminology
+
+- **Fully Qualified Namespace Name** is used to refer to the host name assigned to an Event Hubs namespace as part of the provisioning process.  It typically has the form `{{ Namespace Name }}.servicebus.windows.net`.
+
+- **Event Hubs Host Name** is a synonym for the fully qualified namespace name.  It is a legacy term that sometimes appears in the Azure portal.
+
+- **Custom Endpoint Address** is used to refer to a developer-provided endpoint address which is resolvable in their environment and which should be used when establishing the connection to the Event Hubs service.  It will typically have the form `sb://{{ custom host name }}`.
+
+## Why this is needed
+
+Some of our enterprise customers with environments using unconventional proxy configurations or with certain configurations of an Express Route circuit are unable to connect from their on-premises network to the Event Hubs service.  While it is expected that the [Event Hubs Private Link](https://docs.microsoft.com/en-us/azure/event-hubs/private-link-service) feature, currently in preview, will unblock the Express Route scenario.  However, Private Link is a premium feature with cost considerations and which carries a set of [limitations](https://docs.microsoft.com/en-us/azure/private-link/private-link-service-overview#limitations).  It is also not certain that the Private Link feature will resolve all blocking issues, such as those related to local proxying.
+
+## High level scenarios
+
+### Connecting on a restricted network
+
+A local hospital employs a strict set of network policies to ensure the safety of their network and the security of their patient information.  As part of this strategy, all traffic is routed through a security proxy which performs a threat analysis and logging of connections.
+
+In order to allow for trusted internal applications to embrace cloud services, a special bank of IP addresses has been reserved which passes through a secure DMZ path rather than the security proxy.  In order for trusted connections to work appropriately, they must resolve to one of the reserved IP addresses using a CNAME configured in the local DNS service.
+
+Because of these restrictions, applications are unable to perform direct connections to the Azure Event Hubs service using the standard endpoint address and require the ability to specify a custom host name to ensure they route through the proper intermediary for the connection to be made.
+
+## Proposed approach
+
+- Extend the `EventHubConnectionOptions` to allow a custom endpoint address to be specified.
+
+- When establishing the connection to the Event Hubs service, use the custom endpoint address if it was specified.  If not, follow the current logic for determining and using the standard endpoint address.
+
+- Once the connection is established, ignore the custom endpoint address in favor of the standard endpoint address when sending the AMQP open frame and for authorization purposes.
```

Comment:
```suggestion
- Once the connection is established, prefer the custom endpoint address in favor of the standard endpoint address when sending the AMQP open frame and for authorization purposes.
```

Reply:
Ignore is correct here;  it gets used for the connection and then the standard endpoint is used for service communication.   😄